### DISCUSSION TITLE
Animations!

### DISCUSSION THREAD

**Discussion by BadDogSkyrim** [BadDogSkyrim/PyNifly#155] [Ideas]
I now have animation import & export working in Pynifly but since I don't do animations there's a lot I don't know about how this ought to work.

Right now I'm looking at the relationship between keyframes and the actual exported animation frames. I've been exporting keyframes, but it seems that the game doesn't do smooth interpolation between frames so that might look jerky. Or I can export every frame, but I assume at some point the game engine blows up. Or I can choose a number, like 30 fps, or let the user choose it, and just export that many frames.

What do people expect?

**Comment by Shiagur** [comment]
I think that all animations in Fallout 4 are at 30fps and might have all frames baked in, but I'm not sure. I've imported first-person animations into Blender (after multiple converts) and looked at 3dsmax videos where the HKX files were imported directly and they too seemed to have keyframes at every frame. 
Also, do things like inverse kinematics and bone constraints work when exporting? I know they get baked in when exporting to FBX, maybe you can do something like that?
Also kudos for adding the animation stuff in the first place. I hope we can make all kinds of animations in the future with this, but no pressure.

Edit: Could you also add support for annotations, like SoundPlay and reloadComplete, etc.?

**Comment by dagobaking** [comment]
Can you unpack a vanilla hkx to see which frame-rate and structure they use?

I am also pretty sure its 30FPS.

I agree with Shiagur that annotations would be very helpful.

Also, even a basic tutorial. Or, even better, a set of demo proof of concept files showing how it works would be very helpful. I tried to test with some vanilla files and skeleton. But, I got errors and could not complete a full import/export of hkx.

**Comment by BadDogSkyrim** [comment]
Annotations work, they show up as markers attached to the scene. All the animations I've looked at have 30 frames per second. I tried exporting with a gap and there was no interpolation--the animation just jumped between positions.

Export is done by evaluating the curves at each 30fps mark. If you're using a rig to drive the armature you do have to bake the animation on the armature so there's something to evaluate. 

Right now I'm working on an new idle animation for a werewolf with a new skeleton. Once that's working I'll maybe post some tutorial videos.

**Reply by dagobaking** [reply]
Great! Thank you.

Side note: do the annotations also work the same way on export?

**Comment by BadDogSkyrim** [comment]
If you're willing, send me any Blender animation files you're working on. I need examples of what people do so I can make sure it's covered.

**Reply by Shiagur** [reply]
sure, I can do that. How do you want to receive it? Maybe Discord so I can explain a little about what I'm doing?

**Reply by dagobaking** [reply]
I don't have any specific project file. I was just picking a vanilla hkx idle at random and importing onto the zex skeleton. Trying to do a full cycle of importing and then exporting successfully.

**Comment by BadDogSkyrim** [comment]
Sure, whatever is convenient. I hang out on Ousnius' discord, and here https://discord.gg/QyvgakC5 , or choose your own.

**Comment by dagobaking** [comment]
> Sure, whatever is convenient. I hang out on Ousnius' discord, and here https://discord.gg/QyvgakC5 , or choose your own.

I checked in at those discords a few times. But, didn't see you there.

We have an animation channel at the AAF discord where I think your input would be helpful to other animators in FO4: https://discord.gg/wzTEXbM

If you're interested I can set you up with perms to reach the channel, etc.

**Comment by BadDogSkyrim** [comment]
Ya, I'm there. Not sure I can be a huge help. I really don't know enough about animations.
